cmd/outline-ss-server/config.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2024 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"os"
+
+	"gopkg.in/yaml.v2"
+)
+
+type Service struct {
+	Listeners []Listener
+	Keys      []Key
+}
+
+type Listener struct {
+	Type    string
+	Address string
+}
+
+type Key struct {
+	ID     string
+	Cipher string
+	Secret string
+}
+
+type Config struct {
+	Services []Service
+
+	// Deprecated: Keys exists for historical compatibility. This is ignored if top-level `services` is specified.
+	Keys []struct {
+		ID     string
+		Port   int
+		Cipher string
+		Secret string
+	}
+}
+
+// Reads a config from a filename and parses it as a [Config].
+func ReadConfig(filename string) (*Config, error) {
+	config := Config{}
+	configData, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read config: %w", err)
+	}
+	err = yaml.Unmarshal(configData, &config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse config: %w", err)
+	}
+	if config.Services == nil {
+		// This is a deprecated config format. We need to transform it to to the new format.
```

Comment:
I'm sorry for the question, but still.

We also have an option when there is no service or keys in the configuration file. Why is this option not being tested?